File: 1031-maximum-sum-of-two-non-overlapping-subarrays/1031-maximum-sum-of-two-non-overlapping-subarrays.java
class Solution {
    
    public int maxSum(int []prefix,int M,int N){
        int ans=0,ml=0;
        for(int i=M+N;i<prefix.length;i++){
            ml=Math.max(ml,prefix[i-M]-prefix[i-M-N]);
            ans=Math.max(ans,ml+prefix[i]-prefix[i-M]);
        }
        return ans;
    }
    
    public int maxSumTwoNoOverlap(int[] nums, int firstLen, int secondLen) {
        int []prefix=new int[nums.length+1];
        for(int i=0;i<nums.length;i++){
            prefix[i+1]=prefix[i]+nums[i];
        }
        
        return Math.max(maxSum(prefix,firstLen,secondLen),maxSum(prefix,secondLen,firstLen));
        
    }
}

File: 1031-maximum-sum-of-two-non-overlapping-subarrays/1031-maximum-sum-of-two-non-overlapping-subarrays.cpp
class Solution {
public:
    
    int maxSub(vector<int>&prefix,int m,int n){

        int ans=0,ml=0;
        for(int i=m+n;i<prefix.size();i++){
            ml=max(ml,prefix[i-n]-prefix[i-n-m]);
            ans=max(ans,ml+prefix[i]-prefix[i-n]);
        }
        return ans;
    }
    
    int maxSumTwoNoOverlap(vector<int>& nums, int firstLen, int secondLen) {
        
        vector<int>prefix(nums.size()+1,0);
        for(int i=0;i<nums.size();i++){
            prefix[i+1]=prefix[i]+nums[i];
        }
        return max(maxSub(prefix,firstLen,secondLen),maxSub(prefix,secondLen,firstLen));
    }
};